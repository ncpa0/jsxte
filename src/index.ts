import "./utilities/array-flat-polyfill";

export * from "./express/index";
export * from "./jsx/jsx.types";

export { defineContext } from "./component-api/component-api";
export { DomRenderer } from "./dom-renderer/dom-renderer";
export { ErrorBoundary } from "./error-boundary/error-boundary";
export {
  renderToHtml,
  renderToHtmlAsync,
} from "./html-renderer/render-to-html";
export {
  renderToJson,
  renderToJsonAsync,
} from "./json-renderer/render-to-json";
export { createElement } from "./jsx/jsx-runtime";
export { JsxteRenderError } from "./jsxte-render-error";
export { JsxteRenderer } from "./renderer/renderer";
export { memo } from "./utilities/memo";

export type {
  ComponentApi,
  ContextDefinition,
} from "./component-api/component-api";
export type { DomRenderOptions } from "./dom-renderer/dom-renderer";
export type { HtmlRenderOptions } from "./html-renderer/render-to-html";
export type { JsxteJson } from "./json-renderer/jsx-elem-to-json";
export type { JsonRenderOptions } from "./json-renderer/render-to-json";
export type { AttributeBool, HTMLProps } from "./jsx/base-html-tag-props";
export type { InputType } from "./jsx/prop-types/input-jsx-props";
export type { Crossorigin } from "./jsx/prop-types/shared/crossorigin";
export type { RefererPolicy } from "./jsx/prop-types/shared/referer-policy";
export type { Target } from "./jsx/prop-types/shared/target";
export type { ElementGenerator, RendererOptions } from "./renderer/renderer";
