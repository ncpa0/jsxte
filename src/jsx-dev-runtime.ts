export * from "./jsx/jsx-dev-runtime";
