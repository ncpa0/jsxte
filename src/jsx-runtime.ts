export * from "./jsx/jsx.types";
export * from "./jsx/jsx-runtime";
