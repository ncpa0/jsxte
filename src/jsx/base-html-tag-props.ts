import type { Rewrap } from "../html-parser/types";

export type AttributeBool = true | false | "true" | "false";

export type HTMLProps<T extends object = never> = Rewrap<
  ExtendBaseProps<
    [T] extends [never]
      ? JSXTE.BaseHTMLTagProps
      : Partial<T> & JSXTE.BaseHTMLTagProps
  >
>;

type ExtendBaseProps<P> = {
  [K in keyof P]: JSXTE.AttributeAcceptedTypes extends {
    [E in K]: infer T;
  }
    ? T | P[K]
    : P[K];
};

declare global {
  namespace JSXTE {
    interface AttributeAcceptedTypes {}

    export type TagElement = {
      type: "tag";
      tag:
        | string
        | ((props: ElementProps) => Element)
        | ((props: ElementProps) => Promise<Element>);
      props: ElementProps;
    };

    export type TextNodeElement = {
      type: "textNode";
      text: string;
    };

    export type SyncElement = TagElement | TextNodeElement;

    type ElementChildren =
      | JSX.Element
      | string
      | number
      | Array<
          JSX.Element | string | number | Array<JSX.Element | string | number>
        >;

    type ElementProps = {
      children?: ElementChildren;
      [k: string]: any;
    };

    type PropsWithChildren<P extends object> = P & {
      children?: JSXTE.ElementChildren;
    };

    type Component<P extends object = {}> = (
      props: PropsWithChildren<P>
    ) => JSX.Element;

    type AsyncComponent<P extends object = {}> = (
      props: PropsWithChildren<P>
    ) => Promise<JSX.Element>;

    export interface BaseHTMLTagProps {
      children?: ElementChildren;

      accesskey?: string;
      class?: string;
      contenteditable?: AttributeBool;
      dir?: "ltr" | "rtl" | "auto";
      draggable?: AttributeBool | "auto";
      hidden?: AttributeBool;
      id?: string;
      inert?: AttributeBool;
      lang?: string;
      onabort?: string;
      onafterprint?: string;
      onbeforeprint?: string;
      onbeforeunload?: string;
      onblur?: string;
      oncanplay?: string;
      oncanplaythrough?: string;
      onchange?: string;
      onclick?: string;
      oncontextmenu?: string;
      oncopy?: string;
      oncuechange?: string;
      oncut?: string;
      ondblclick?: string;
      ondrag?: string;
      ondragend?: string;
      ondragenter?: string;
      ondragleave?: string;
      ondragover?: string;
      ondragstart?: string;
      ondrop?: string;
      ondurationchange?: string;
      onemptied?: string;
      onended?: string;
      onerror?: string;
      onfocus?: string;
      onhashchange?: string;
      oninput?: string;
      oninvalid?: string;
      onkeydown?: string;
      onkeypress?: string;
      onkeyup?: string;
      onload?: string;
      onloadeddata?: string;
      onloadedmetadata?: string;
      onloadstart?: string;
      onmousedown?: string;
      onmousemove?: string;
      onmouseout?: string;
      onmouseover?: string;
      onmouseup?: string;
      onmousewheel?: string;
      onoffline?: string;
      ononline?: string;
      onpagehide?: string;
      onpageshow?: string;
      onpaste?: string;
      onpause?: string;
      onplay?: string;
      onplaying?: string;
      onpopstate?: string;
      onprogress?: string;
      onratechange?: string;
      onreset?: string;
      onresize?: string;
      onscroll?: string;
      onsearch?: string;
      onseeked?: string;
      onseeking?: string;
      onselect?: string;
      onstalled?: string;
      onstorage?: string;
      onsubmit?: string;
      onsuspend?: string;
      ontimeupdate?: string;
      ontoggle?: string;
      onunload?: string;
      onvolumechange?: string;
      onwaiting?: string;
      onwheel?: string;
      role?: string;
      slot?: string;
      spellcheck?: AttributeBool;
      style?: string;
      tabindex?: string | number;
      title?: string;
      translate?: "yes" | "no";
    }
  }
}
