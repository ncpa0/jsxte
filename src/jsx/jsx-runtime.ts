import type { ComponentApi } from "../component-api/component-api";

type CreateElementProps = {
  [k: string]: any;
  children?: JSXTE.ElementChildren;
};

(() => {
  if(typeof Symbol.toHtmlTag === "symbol") return;

  Object.defineProperty(Symbol, "toHtmlTag", {
    value: Symbol("toHtmlTag"),
    enumerable: false,
    configurable: true,
  });
})();

declare global {
  interface SymbolConstructor {
    /**
     * Method under this property defines how an object should be converted to
     * HTML by the JSXTE renderer.
     *
     * @example
     *   class User {
     *     constructor(
     *       public firstname: string,
     *       public lastname: string,
     *       public email: string,
     *       public age: number,
     *     ) {}
     *
     *     [Symbol.toHtmlTag]() {
     *       return `${this.firstname} ${this.lastname}`;
     *     }
     *   }
     *
     *   const user = new User("John", "Doe", "joedoe@gmai.com", 26);
     *   renderToHtml(<div>{user}</div>); // => <div>John Doe</div>
     */
    readonly toHtmlTag: symbol;
  }
}

const mapChildren = (
  children: JSXTE.ElementChildren,
  accumulator: JSX.Element[],
) => {
  // eslint-disable-next-line @typescript-eslint/switch-exhaustiveness-check
  switch (typeof children) {
    case "string":
      accumulator.push({ type: "textNode", text: children });
      break;
    case "number":
      accumulator.push({ type: "textNode", text: children.toString() });
      break;
    case "object":
      if (Array.isArray(children)) {
        for (let i = 0; i < children.length; i++) {
          const child = children[i]!;
          mapChildren(child, accumulator);
        }
      } else if (children != null) {
        if (
          Symbol.toHtmlTag in children &&
          typeof children[Symbol.toHtmlTag] === "function"
        ) {
          const html = String((children[Symbol.toHtmlTag] as Function)());
          accumulator.push({ type: "textNode", text: html });
        } else {
          accumulator.push(children);
        }
      }
      break;
  }

  return accumulator;
};

export const createElement = (
  tag:
    | string
    | ((props: any, contextMap: ComponentApi) => JSX.Element)
    | ((props: any, contextMap: ComponentApi) => Promise<JSX.Element>),
  props?: CreateElementProps,
  ...children: Array<
    JSX.Element | string | number | Array<JSX.Element | string | number>
  >
): JSX.Element => {
  props ??= {};

  const finalChildren: JSX.Element[] = [];

  for (let i = 0; i < children.length; i++) {
    mapChildren(children[i]!, finalChildren);
  }

  if (props?.children) {
    mapChildren(props.children, finalChildren);
  }

  props.children = finalChildren;

  Object.freeze(finalChildren);
  Object.freeze(props);

  return {
    type: "tag",
    // @ts-expect-error
    tag,
    props: props as JSXTE.ElementProps,
  };
};

export const jsx = createElement;
export const jsxs = jsx;
export const _jsx = jsx;
export const _jsxs = jsx;

export const Fragment = "";
export const _Fragment = Fragment;
