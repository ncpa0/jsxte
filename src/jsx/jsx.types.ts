import type { HTMLProps } from "./base-html-tag-props";
import type { AnchorTagProps } from "./prop-types/a-jsx-props";
import type { AreaTagProps } from "./prop-types/area-jsx-props";
import type { AudioTagProps } from "./prop-types/audio-jsx-props";
import type { BaseTagProps } from "./prop-types/base-jsx-props";
import type { BdoTagProps } from "./prop-types/bdo-jsx-props";
import type { BlockquoteTagProps } from "./prop-types/blockquote-jsx-props";
import type { ButtonTagProps } from "./prop-types/button-jsx-props";
import type { CanvasTagProps } from "./prop-types/canvas-jsx-props";
import type { ColTagProps } from "./prop-types/col-jsx-props";
import type { ColgroupTagProps } from "./prop-types/colgroup-jsx-props";
import type { DataTagProps } from "./prop-types/data-jsx-props";
import type { DelTagProps } from "./prop-types/del-jsx-props";
import type { DetailsTagProps } from "./prop-types/details-jsx-props";
import type { DialogTagProps } from "./prop-types/dialog-jsx-props";
import type { EmbedTagProps } from "./prop-types/embed-jsx-props";
import type { FieldsetTagProps } from "./prop-types/fieldset-jsx-prop";
import type { FormTagProps } from "./prop-types/form-jsx-props";
import type { HtmlTagProps } from "./prop-types/html-jsx-props";
import type { IframeTagProps } from "./prop-types/iframe-jsx-props";
import type { ImgTagProps } from "./prop-types/img-jsx-props";
import type { InputTagProps } from "./prop-types/input-jsx-props";
import type { InsTagProps } from "./prop-types/ins-jsx-props";
import type { LabelTagProps } from "./prop-types/label-jsx-props";
import type { LiTagProps } from "./prop-types/li-jsx.props";
import type { LinkTagProps } from "./prop-types/link-jsx-props";
import type { MapTagProps } from "./prop-types/map-jsx-props";
import type { MetaTagProps } from "./prop-types/meta-jsx-props";
import type { MeterTagProps } from "./prop-types/meter-jsx-props";
import type { ObjectTagProps } from "./prop-types/object-jsx-props";
import type { OlTagProps } from "./prop-types/ol-jsx-props";
import type { OptgroupTagProps } from "./prop-types/optgroup-jsx-props";
import type { OptionTagProps } from "./prop-types/option-jsx-props";
import type { OutputTagProps } from "./prop-types/output-jsx-props";
import type { ParamTagProps } from "./prop-types/param-jsx-props";
import type { ProgressTagProps } from "./prop-types/progress-jsx-props";
import type { QTagProps } from "./prop-types/q-jsx-props";
import type { ScriptTagProps } from "./prop-types/script-jsx-props";
import type { SelectTagProps } from "./prop-types/select-jsx-props";
import type { SourceTagProps } from "./prop-types/source-jsx-props";
import type { StyleTagProps } from "./prop-types/style-jsx-props";
import type { TdTagProps } from "./prop-types/td-jsx-props";
import type { TextareaTagProps } from "./prop-types/textarea-jsx-props";
import type { ThTagProps } from "./prop-types/th-jsx-props";
import type { TimeTagProps } from "./prop-types/time-jsx-props";
import type { TrackTagProps } from "./prop-types/track-jsx-props";
import type { VideoTagProps } from "./prop-types/video-jsx-props";

declare global {
  namespace JSX {
    type AsyncElement = Promise<JSXTE.TagElement | JSXTE.TextNodeElement>;
    type SyncElement = JSXTE.TagElement | JSXTE.TextNodeElement;

    type Element = SyncElement | AsyncElement;

    interface IntrinsicElements {
      a: HTMLProps<AnchorTagProps>;
      abbr: HTMLProps;
      address: HTMLProps;
      area: HTMLProps<AreaTagProps>;
      article: HTMLProps;
      aside: HTMLProps;
      audio: HTMLProps<AudioTagProps>;
      b: HTMLProps;
      base: HTMLProps<BaseTagProps>;
      bdi: HTMLProps;
      bdo: HTMLProps<BdoTagProps>;
      big: HTMLProps;
      blockquote: HTMLProps<BlockquoteTagProps>;
      body: HTMLProps;
      br: HTMLProps;
      button: HTMLProps<ButtonTagProps>;
      canvas: HTMLProps<CanvasTagProps>;
      caption: HTMLProps;
      cite: HTMLProps;
      code: HTMLProps;
      col: HTMLProps<ColTagProps>;
      colgroup: HTMLProps<ColgroupTagProps>;
      data: HTMLProps<DataTagProps>;
      datalist: HTMLProps;
      dd: HTMLProps;
      del: HTMLProps<DelTagProps>;
      details: HTMLProps<DetailsTagProps>;
      dfn: HTMLProps;
      dialog: HTMLProps<DialogTagProps>;
      div: HTMLProps;
      dl: HTMLProps;
      dt: HTMLProps;
      em: HTMLProps;
      embed: HTMLProps<EmbedTagProps>;
      fieldset: HTMLProps<FieldsetTagProps>;
      figcaption: HTMLProps;
      figure: HTMLProps;
      footer: HTMLProps;
      form: HTMLProps<FormTagProps>;
      h1: HTMLProps;
      h2: HTMLProps;
      h3: HTMLProps;
      h4: HTMLProps;
      h5: HTMLProps;
      h6: HTMLProps;
      head: HTMLProps;
      header: HTMLProps;
      hgroup: HTMLProps;
      hr: HTMLProps;
      html: HTMLProps<HtmlTagProps>;
      i: HTMLProps;
      iframe: HTMLProps<IframeTagProps>;
      img: HTMLProps<ImgTagProps>;
      input: HTMLProps<InputTagProps>;
      ins: HTMLProps<InsTagProps>;
      kbd: HTMLProps;
      keygen: HTMLProps;
      label: HTMLProps<LabelTagProps>;
      legend: HTMLProps;
      li: HTMLProps<LiTagProps>;
      link: HTMLProps<LinkTagProps>;
      main: HTMLProps;
      map: HTMLProps<MapTagProps>;
      mark: HTMLProps;
      menu: HTMLProps;
      menuitem: HTMLProps;
      meta: HTMLProps<MetaTagProps>;
      meter: HTMLProps<MeterTagProps>;
      nav: HTMLProps;
      noindex: HTMLProps;
      noscript: HTMLProps;
      object: HTMLProps<ObjectTagProps>;
      ol: HTMLProps<OlTagProps>;
      optgroup: HTMLProps<OptgroupTagProps>;
      option: HTMLProps<OptionTagProps>;
      output: HTMLProps<OutputTagProps>;
      p: HTMLProps;
      param: HTMLProps<ParamTagProps>;
      picture: HTMLProps;
      pre: HTMLProps;
      progress: HTMLProps<ProgressTagProps>;
      q: HTMLProps<QTagProps>;
      rp: HTMLProps;
      rt: HTMLProps;
      ruby: HTMLProps;
      s: HTMLProps;
      samp: HTMLProps;
      slot: HTMLProps;
      script: HTMLProps<ScriptTagProps>;
      section: HTMLProps;
      select: HTMLProps<SelectTagProps>;
      small: HTMLProps;
      source: HTMLProps<SourceTagProps>;
      span: HTMLProps;
      strong: HTMLProps;
      style: HTMLProps<StyleTagProps>;
      sub: HTMLProps;
      summary: HTMLProps;
      sup: HTMLProps;
      table: HTMLProps;
      template: HTMLProps;
      tbody: HTMLProps;
      td: HTMLProps<TdTagProps>;
      textarea: HTMLProps<TextareaTagProps>;
      tfoot: HTMLProps;
      th: HTMLProps<ThTagProps>;
      thead: HTMLProps;
      time: HTMLProps<TimeTagProps>;
      title: HTMLProps;
      tr: HTMLProps;
      track: HTMLProps<TrackTagProps>;
      u: HTMLProps;
      ul: HTMLProps;
      var: HTMLProps;
      video: HTMLProps<VideoTagProps>;
      wbr: HTMLProps;
      webview: HTMLProps;

      // SVG
      svg: HTMLProps;

      // web components
    }
  }
}
