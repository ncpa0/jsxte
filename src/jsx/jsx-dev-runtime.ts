import type { ComponentApi } from "../component-api/component-api";
import type { CreateElementProps } from "./jsx-runtime";
import { Fragment, jsx, jsxs } from "./jsx-runtime";

const jsxDEV = (
  tag:
    | string
    | ((props: any, contextMap: ComponentApi) => JSX.Element)
    | ((props: any, contextMap: ComponentApi) => Promise<JSX.Element>),
  props?: CreateElementProps,
) => {
  const { children, ...restProps } = props ?? {};
  if (Array.isArray(children)) {
    return jsx(tag, restProps, ...children);
  }
  return jsx(tag, restProps, children);
};

export { Fragment, jsx, jsxs, jsxDEV };
