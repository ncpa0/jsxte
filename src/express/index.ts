import { renderToHtmlAsync } from "../html-renderer/render-to-html";
import { createElement } from "../jsx/jsx-runtime";

/**
 * Express.js support.
 *
 * Allows to support Express.js builtin templating.
 */
const __express = async <P extends object>(
  filePath: string,
  options: P,
  callback: (e: any, rendered?: string | undefined) => void,
) => {
  // @ts-expect-error
  options ??= {};

  try {
    // eslint-disable-next-line
    const Component: JSXTE.AsyncComponent<P> = require(filePath).default;
    const html = await renderToHtmlAsync(createElement(Component, options));
    return callback(null, html);
  } catch (e) {
    callback(e);
  }
};

type ExpressApp = {
  engine(
    ext: string,
    fn: (
      path: string,
      options: object,
      callback: (e: any, rendered?: string) => void,
    ) => void,
  ): any;
  set(setting: string, val: any): any;
};

/**
 * Extends the Express server to use jsxte engine for rendering views.
 *
 * View files should have `.js` extension and have a default export with a jsx
 * component.
 */
export const expressExtend = (app: ExpressApp) => {
  app.engine("js", __express);
  app.set("view engine", "js");
};
