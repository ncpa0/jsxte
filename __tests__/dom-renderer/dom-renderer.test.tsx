import { JSDOM } from "jsdom";
import { describe, expect, it } from "vitest";
import { DomRenderer } from "../../src/index";
// @ts-ignore
import { Fragment, jsx } from "../../src/jsx/jsx-runtime";

describe("DomRenderer", () => {
  const { window } = new JSDOM();
  const renderer = new DomRenderer(window);

  it("renders an empty div", () => {
    const element = renderer.render(<div />) as HTMLDivElement;
    expect(element).toMatchObject(window.document.createElement("div"));
    expect(element.outerHTML).toMatchSnapshot();
  });

  it("renders a div with text", () => {
    const element = renderer.render(<div>Hello</div>) as HTMLDivElement;
    const expected = window.document.createElement("div");
    expected.textContent = "Hello";
    expect(element).toMatchObject(expected);
    expect(element.outerHTML).toMatchSnapshot();
  });

  it("renders a div with a child", () => {
    const element = renderer.render(
      <div>
        <span>Hello</span>
      </div>,
    ) as HTMLDivElement;
    const expected = window.document.createElement("div");
    const span = window.document.createElement("span");
    span.textContent = "Hello";
    expected.appendChild(span);
    expect(element).toMatchObject(expected);
    expect(element.outerHTML).toMatchSnapshot();
  });

  it("renders a div with a child and a class", () => {
    const element = renderer.render(
      <div class="foo">
        <span>Hello</span>
      </div>,
    ) as HTMLDivElement;
    const expected = window.document.createElement("div");
    expected.className = "foo";
    const span = window.document.createElement("span");
    span.textContent = "Hello";
    expected.appendChild(span);
    expect(element).toMatchObject(expected);
    expect(element.outerHTML).toMatchSnapshot();
  });

  it("renders a nested structure with attributes", () => {
    const element = renderer.render(
      <body id="root" class="foo" draggable>
        <head>
          <nav>
            <ul>
              <li>
                <a href="/">Home</a>
              </li>
              <li>
                <a href="/about">About</a>
              </li>
            </ul>
          </nav>
        </head>
        <main class="main-contents" style="padding: 1rem; margin: 1rem;">
          <div>
            <h1>Todo's:</h1>
            {"- Write some code"}
            <br />
            {"- Write some tests"}
            <br />
            {"- Profit"}
          </div>
          <div class="interpolated-str">
            This {"is"} a {"string"} with {"interpolated"} values
          </div>
        </main>
      </body>,
    ) as HTMLElement;

    expect(element).toMatchSnapshot();
  });

  describe("custom window type", () => {
    class HtmlElem {
      static make(
        tagName: string,
        attr: Record<string, string>,
        ...children: Array<HtmlElem | HtmlFrag | HtmlTxt>
      ) {
        const elem = new HtmlElem(tagName);
        for (const [name, value] of Object.entries(attr)) {
          elem.setAttribute(name, value);
        }
        for (const child of children) {
          elem.appendChild(child);
        }
        return elem;
      }

      public children: Array<HtmlElem | HtmlFrag | HtmlTxt> = [];
      public attributes: Record<string, string> = {};
      constructor(public tagName: string) {}
      appendChild<T extends HtmlElem | HtmlFrag | HtmlTxt>(child: T): T {
        this.children.push(child);
        return child;
      }
      setAttribute(name: string, value: string) {
        this.attributes[name] = value;
      }
    }

    class HtmlFrag {
      static make(
        ...children: Array<HtmlElem | HtmlFrag | HtmlTxt>
      ) {
        const elem = new HtmlFrag();
        for (const child of children) {
          elem.appendChild(child);
        }
        return elem;
      }

      public children: Array<HtmlElem | HtmlFrag | HtmlTxt> = [];
      appendChild<T extends HtmlElem | HtmlFrag | HtmlTxt>(child: T): T {
        this.children.push(child);
        return child;
      }
    }

    class HtmlTxt {
      public textContent: string | null = null;
      constructor(public text: string) {}
    }

    class Doc {
      public createElement(tagName: string) {
        return new HtmlElem(tagName);
      }

      public createTextNode(text: string) {
        return new HtmlTxt(text);
      }

      public createDocumentFragment() {
        return new HtmlFrag();
      }
    }

    class Win {
      public document = new Doc();
    }

    const window = new Win();

    it("renders a nested structure with attributes", () => {
      const renderer = new DomRenderer(window);

      const elem = renderer.render(
        <div id="root" class="foo" hidden>
          <span>Hello</span>
          <div contenteditable={false}>
            <ul class="links-list unordered column" role="navigation">
              <li>
                <a href="/">Home</a>
              </li>
              <li>
                <a href="/about">About</a>
              </li>
            </ul>
          </div>
        </div>,
      );

      expect(elem).toMatchObject(
        HtmlElem.make(
          "div",
          { id: "root", class: "foo", hidden: "hidden" },
          HtmlElem.make("span", {}, new HtmlTxt("Hello")),
          HtmlElem.make(
            "div",
            {},
            HtmlElem.make(
              "ul",
              {
                class: "links-list unordered column",
                role: "navigation",
              },
              HtmlElem.make(
                "li",
                {},
                HtmlElem.make("a", { href: "/" }, new HtmlTxt("Home")),
              ),
              HtmlElem.make(
                "li",
                {},
                HtmlElem.make("a", { href: "/about" }, new HtmlTxt("About")),
              ),
            ),
          ),
        ),
      );
    });
  });
});
