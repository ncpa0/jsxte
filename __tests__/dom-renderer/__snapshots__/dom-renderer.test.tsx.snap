// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DomRenderer > renders a div with a child 1`] = `"<div><span>Hello</span></div>"`;

exports[`DomRenderer > renders a div with a child and a class 1`] = `"<div class="foo"><span>Hello</span></div>"`;

exports[`DomRenderer > renders a div with text 1`] = `"<div>Hello</div>"`;

exports[`DomRenderer > renders a nested structure with attributes 1`] = `
<body
  class="foo"
  draggable="draggable"
  id="root"
>
  <head>
    <nav>
      <ul>
        <li>
          <a
            href="/"
          >
            Home
          </a>
        </li>
        <li>
          <a
            href="/about"
          >
            About
          </a>
        </li>
      </ul>
    </nav>
  </head>
  <main
    class="main-contents"
    style="padding: 1rem; margin: 1rem;"
  >
    <div>
      <h1>
        Todo's:
      </h1>
      - Write some code
      <br />
      - Write some tests
      <br />
      - Profit
    </div>
    <div
      class="interpolated-str"
    >
      This 
      is
       a 
      string
       with 
      interpolated
       values
    </div>
  </main>
</body>
`;

exports[`DomRenderer > renders an empty div 1`] = `"<div></div>"`;
